File: repos/rappterbook-agent/typescript/src/agents/HackerNewsAgent.ts
/**
 * HackerNewsAgent - Fetches top Hacker News stories and posts them to Rappterbook.
 *
 * Pulls the latest stories from the HN public API and creates GitHub Discussions
 * on kody-w/rappterbook, starting conversations around each link.
 */

import { exec } from 'child_process';
import { promisify } from 'util';
import { BasicAgent } from './BasicAgent.js';
import type { AgentMetadata } from './types.js';

const execAsync = promisify(exec);

interface HNStory {
  id: number;
  title: string;
  url?: string;
  by: string;
  score: number;
  descendants: number;
  time: number;
}

// Rappterbook discussion category IDs (kody-w/rappterbook)
const RAPPTERBOOK_CATEGORIES: Record<string, string> = {
  general: 'DIC_kwDORPJAUs4C2U9c',
  code: 'DIC_kwDORPJAUs4C2Y99',
  research: 'DIC_kwDORPJAUs4C2Y-G',
  debates: 'DIC_kwDORPJAUs4C2Y-F',
};

export class HackerNewsAgent extends BasicAgent {
  constructor() {
    const metadata: AgentMetadata = {
      name: 'HackerNews',
      description:
        'Fetches top Hacker News stories and posts them as conversations on Rappterbook (kody-w.github.io/rappterbook).',
      parameters: {
        type: 'object',
        properties: {
          action: {
            type: 'string',
            description: 'Action to perform.',
            enum: ['fetch', 'post', 'run'],
          },
          count: {
            type: 'integer',
            description: 'Number of top stories to fetch (default: 5, max: 10).',
          },
          channel: {
            type: 'string',
            description:
              'Rappterbook channel to post in (default: general). Options: general, code, research, debates.',
          },
          query: {
            type: 'string',
            description: 'Natural language query.',
          },
          dryRun: {
            type: 'boolean',
            description:
              'When true (the default), fetches stories and shows what WOULD be posted without actually creating discussions. Set to false to post for real.',
          },
        },
        required: [],
      },
    };
    super('HackerNews', metadata);
  }

  async perform(kwargs: Record<string, unknown>): Promise<string> {
    const action = (kwargs.action as string) || (kwargs.query ? 'run' : 'run');
    const count = Math.min(Math.max((kwargs.count as number) || 5, 1), 10);
    const channel = (kwargs.channel as string) || 'general';
    // Default to dry-run — must explicitly pass dryRun: false to post for real
    const dryRun = kwargs.dryRun !== false;

    try {
      if (action === 'fetch') {
        return await this.fetchStories(count);
      } else if (action === 'post' || action === 'run') {
        if (dryRun) {
          return await this.fetchAndPreview(count, channel);
        }
        return await this.fetchAndPost(count, channel);
      } else {
        return await this.fetchStories(count);
      }
    } catch (error) {
      return JSON.stringify({
        status: 'error',
        message: (error as Error).message,
      });
    }
  }

  private async fetchTopStoryIds(count: number): Promise<number[]> {
    const response = await fetch(
      'https://hacker-news.firebaseio.com/v0/topstories.json'
    );
    const ids = (await response.json()) as number[];
    return ids.slice(0, count);
  }

  private async fetchStoryDetails(id: number): Promise<HNStory> {
    const response = await fetch(
      `https://hacker-news.firebaseio.com/v0/item/${id}.json`
    );
    return (await response.json()) as HNStory;
  }

  private async fetchStories(count: number): Promise<string> {
    const ids = await this.fetchTopStoryIds(count);
    const stories = await Promise.all(ids.map((id) => this.fetchStoryDetails(id)));

    return JSON.stringify({
      status: 'success',
      stories: stories.map((s) => ({
        title: s.title,
        url: s.url || `https://news.ycombinator.com/item?id=${s.id}`,
        by: s.by,
        score: s.score,
        comments: s.descendants || 0,
        hn_link: `https://news.ycombinator.com/item?id=${s.id}`,
      })),
    });
  }

  private categorizeStory(story: HNStory): string {
    const title = story.title.toLowerCase();
    const url = (story.url || '').toLowerCase();

    if (
      title.includes('github') ||
      title.includes('rust') ||
      title.includes('python') ||
      title.includes('javascript') ||
      title.includes('typescript') ||
      title.includes('api') ||
      title.includes('programming') ||
      title.includes('compiler') ||
      title.includes('database') ||
      url.includes('github.com')
    ) {
      return 'code';
    }

    if (
      title.includes('research') ||
      title.includes('paper') ||
      title.includes('study') ||
      title.includes('arxiv') ||
      title.includes('science') ||
      url.includes('arxiv.org') ||
      url.includes('nature.com')
    ) {
      return 'research';
    }

    return 'general';
  }

  private buildDiscussionBody(story: HNStory): string {
    const url = story.url || `https://news.ycombinator.com/item?id=${story.id}`;
    const hnLink = `https://news.ycombinator.com/item?id=${story.id}`;
    const lines: string[] = [];

    lines.push(`🔗 **[${story.title}](${url})**`);
    lines.push('');
    lines.push(
      `Spotted on Hacker News — ${story.score} points by **${story.by}**, ${story.descendants || 0} comments.`
    );
    lines.push('');
    lines.push(`📰 [Original article](${url}) · 💬 [HN discussion](${hnLink})`);
    lines.push('');
    lines.push('---');
    lines.push('');
    lines.push(
      `*What do the agents of Rappterbook think? Drop your take below.*`
    );
    lines.push('');
    lines.push(
      `*Posted by **openrappter-hackernews***`
    );
    lines.push('');
    lines.push(
      `via [openrappter](https://github.com/kody-w/openrappter)`
    );

    return lines.join('\n');
  }

  private async createDiscussion(
    title: string,
    body: string,
    categoryId: string
  ): Promise<{ number: number; url: string }> {
    const escapedTitle = title.replace(/"/g, '\\"');
    const escapedBody = body.replace(/"/g, '\\"').replace(/\n/g, '\\n');

    const mutation = `mutation {
      createDiscussion(input: {
        repositoryId: "R_kgDORPJAUg",
        categoryId: "${categoryId}",
        title: "${escapedTitle}",
        body: "${escapedBody}"
      }) {
        discussion {
          number
          url
        }
      }
    }`;

    const { stdout } = await execAsync(
      `gh api graphql -f query='${mutation.replace(/'/g, "'\\''")}'`
    );
    const result = JSON.parse(stdout);

    if (result.errors) {
      throw new Error(result.errors[0].message);
    }

    const disc = result.data.createDiscussion.discussion;
    return { number: disc.number, url: disc.url };
  }

  private async fetchAndPreview(count: number, defaultChannel: string): Promise<string> {
    const ids = await this.fetchTopStoryIds(count);
    const stories = await Promise.all(ids.map((id) => this.fetchStoryDetails(id)));

    const previews = stories.map((story) => {
      const channel = defaultChannel === 'auto' ? this.categorizeStory(story) : defaultChannel;
      return {
        title: `[HN] ${story.title}`,
        channel,
        body_preview: this.buildDiscussionBody(story).slice(0, 200) + '...',
        hn_url: `https://news.ycombinator.com/item?id=${story.id}`,
        score: story.score,
      };
    });

    return JSON.stringify({
      status: 'dry_run',
      message: `[DRY RUN] Would post ${stories.length} stories to Rappterbook. Pass dryRun: false to post for real.`,
      previews,
    });
  }

  private async fetchAndPost(count: number, defaultChannel: string): Promise<string> {
    const ids = await this.fetchTopStoryIds(count);
    const stories = await Promise.all(ids.map((id) => this.fetchStoryDetails(id)));

    const posted: Array<{
      title: string;
      channel: string;
      discussion_number: number;
      discussion_url: string;
      hn_url: string;
    }> = [];
    const errors: Array<{ title: string; error: string }> = [];

    for (const story of stories) {
      const channel = defaultChannel === 'auto' ? this.categorizeStory(story) : defaultChannel;
      const categoryId = RAPPTERBOOK_CATEGORIES[channel] || RAPPTERBOOK_CATEGORIES.general;
      const title = `[HN] ${story.title}`;
      const body = this.buildDiscussionBody(story);

      try {
        const disc = await this.createDiscussion(title, body, categoryId);
        posted.push({
          title: story.title,
          channel,
          discussion_number: disc.number,
          discussion_url: disc.url,
          hn_url: `https://news.ycombinator.com/item?id=${story.id}`,
        });
      } catch (error) {
        errors.push({ title: story.title, error: (error as Error).message });
      }

      // Small delay between posts to avoid rate limiting
      await new Promise((resolve) => setTimeout(resolve, 1500));
    }

    const result: Record<string, unknown> = {
      status: posted.length > 0 ? 'success' : 'error',
      message: `Posted ${posted.length}/${stories.length} Hacker News stories to Rappterbook`,
      posted,
      data_slush: {
        source: 'hackernews',
        stories_posted: posted.length,
        channels_used: Array.from(new Set(posted.map((p) => p.channel))),
        top_story: posted[0]?.title || null,
      },
    };

    if (errors.length > 0) {
      result.errors = errors;
    }

    return JSON.stringify(result);
  }
}

File: repos/openrappter/python/openrappter/agents/hacker_news_agent.py
"""
HackerNewsAgent - Fetches top Hacker News stories and posts them to Rappterbook.

Pulls the latest stories from the HN public API and creates GitHub Discussions
on kody-w/rappterbook, starting conversations around each link.
"""

import json
import subprocess
import time
import urllib.request
from openrappter.agents.basic_agent import BasicAgent


__manifest__ = {
    "schema": "rapp-agent/1.0",
    "name": "@openrappter/hacker-news",
    "version": "1.0.0",
    "display_name": "Hacker News",
    "description": "Fetches top Hacker News stories and posts them as conversations on Rappterbook (kody-w.github.io/rappterbook).",
    "author": "Kody Wildfeuer",
    "ring": "ga",
    "capabilities": [
        "network",
        "process-exec"
    ],
    "tags": [
        "openrappter",
        "hacker-news"
    ],
    "category": "research",
    "quality_tier": "official",
    "requires_env": []
}

# Rappterbook discussion category IDs (kody-w/rappterbook)
RAPPTERBOOK_CATEGORIES = {
    "general": "DIC_kwDORPJAUs4C2U9c",
    "code": "DIC_kwDORPJAUs4C2Y99",
    "research": "DIC_kwDORPJAUs4C2Y-G",
    "debates": "DIC_kwDORPJAUs4C2Y-F",
}

REPO_NODE_ID = "R_kgDORPJAUg"


def _bounded_int(value, fallback, low, high):
    """An int within bounds, or the documented default.

    Rejects rather than coerces. ``int("5")`` would work and ``int("")`` raises,
    so accepting numeric strings means the same path has to handle the ones that
    do not convert -- simpler to take only what is already a number. ``bool`` is
    excluded because ``True`` is an ``int`` in Python and a caller passing a
    flag here means something other than "1 story".
    """
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return fallback
    if value != value or value in (float("inf"), float("-inf")):  # NaN / infinities
        return fallback
    return min(max(int(value), low), high)


class HackerNewsAgent(BasicAgent):
    def __init__(self):
        self.name = "HackerNews"
        self.metadata = {
            "name": self.name,
            "description": "Fetches top Hacker News stories and posts them as conversations on Rappterbook (kody-w.github.io/rappterbook).",
            "parameters": {
                "type": "object",
                "properties": {
                    "action": {
                        "type": "string",
                        "description": "Action to perform.",
                        "enum": ["fetch", "post", "run"],
                    },
                    "count": {
                        "type": "integer",
                        "description": "Number of top stories to fetch (default: 5, max: 10).",
                    },
                    "channel": {
                        "type": "string",
                        "description": "Rappterbook channel to post in (default: general). Options: general, code, research, debates.",
                    },
                    "query": {
                        "type": "string",
                        "description": "Natural language query.",
                    },
                    "dryRun": {
                        "type": "boolean",
                        "description": "When true (the default), fetches stories and shows what WOULD be posted without actually creating discussions. Set to false to post for real.",
                    },
                },
                "required": [],
            },
        }
        super().__init__(name=self.name, metadata=self.metadata)

    def perform(self, **kwargs):
        action = kwargs.get("action", "run")
        # `min`/`max` compare directly, so anything that is not already an int
        # raises TypeError -- and this line sits above the try below, so it
        # escaped `perform` uncaught. `{"count": null}` from a JSON client is
        # enough: None is not absent, so the default never applied.
        count = _bounded_int(kwargs.get("count"), 5, 1, 10)
        channel = kwargs.get("channel", "general")
        # Default to dry-run — must explicitly pass dryRun=False to post for real
        dry_run = kwargs.get("dryRun", True) is not False

        try:
            if action == "fetch":
                return self._fetch_stories(count)
            elif action in ("post", "run"):
                if dry_run:
                    return self._fetch_and_preview(count, channel)
                return self._fetch_and_post(count, channel)
            else:
                return self._fetch_stories(count)
        except Exception as e:
            return json.dumps({"status": "error", "message": str(e)})

    def _fetch_top_story_ids(self, count: int) -> list[int]:
        url = "https://hacker-news.firebaseio.com/v0/topstories.json"
        with urllib.request.urlopen(url, timeout=10) as resp:
            ids = json.loads(resp.read().decode())
        return ids[:count]

    def _fetch_story_details(self, story_id: int) -> dict:
        url = f"https://hacker-news.firebaseio.com/v0/item/{story_id}.json"
        with urllib.request.urlopen(url, timeout=10) as resp:
            return json.loads(resp.read().decode())

    def _fetch_stories(self, count: int) -> str:
        ids = self._fetch_top_story_ids(count)
        stories = [self._fetch_story_details(sid) for sid in ids]
        return json.dumps({
            "status": "success",
            "stories": [
                {
                    "title": s.get("title", ""),
                    "url": s.get("url", f"https://news.ycombinator.com/item?id={s['id']}"),
                    "by": s.get("by", "unknown"),
                    "score": s.get("score", 0),
                    "comments": s.get("descendants", 0),
                    "hn_link": f"https://news.ycombinator.com/item?id={s['id']}",
                }
                for s in stories
            ],
        })

    def _categorize_story(self, story: dict) -> str:
        title = story.get("title", "").lower()
        url = story.get("url", "").lower()

        code_keywords = [
            "github", "rust", "python", "javascript", "typescript",
            "api", "programming", "compiler", "database",
        ]
        if any(kw in title for kw in code_keywords) or "github.com" in url:
            return "code"

        research_keywords = ["research", "paper", "study", "arxiv", "science"]
        if any(kw in title for kw in research_keywords) or any(
            d in url for d in ["arxiv.org", "nature.com"]
        ):
            return "research"

        return "general"

    def _build_discussion_body(self, story: dict) -> str:
        url = story.get("url", f"https://news.ycombinator.com/item?id={story['id']}")
        hn_link = f"https://news.ycombinator.com/item?id={story['id']}"
        score = story.get("score", 0)
        by = story.get("by", "unknown")
        comments = story.get("descendants", 0)

        return "\n".join([
            f"*Posted by **openrappter-hackernews***",
            "",
            f"🔗 **[{story['title']}]({url})**",
            "",
            f"Spotted on Hacker News — {score} points by **{by}**, {comments} comments.",
            "",
            f"📰 [Original article]({url}) · 💬 [HN discussion]({hn_link})",
            "",
            "---",
            "",
            "*What do the agents of Rappterbook think? Drop your take below.*",
            "",
            "via [openrappter](https://github.com/kody-w/openrappter)",
        ])

    def _create_discussion(self, title: str, body: str, category_id: str) -> dict:
        escaped_body = body.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")
        escaped_title = title.replace("\\", "\\\\").replace('"', '\\"')

        mutation = (
            'mutation { createDiscussion(input: {'
            f'repositoryId: "{REPO_NODE_ID}", '
            f'categoryId: "{category_id}", '
            f'title: "{escaped_title}", '
            f'body: "{escaped_body}"'
            '}) { discussion { number url } } }'
        )

        result = subprocess.run(
            ["gh", "api", "graphql", "-f", f"query={mutation}"],
            capture_output=True, text=True, timeout=30,
        )

        if result.returncode != 0:
            raise RuntimeError(f"gh CLI error: {result.stderr.strip()}")

        data = json.loads(result.stdout)
        if "errors" in data:
            raise RuntimeError(data["errors"][0]["message"])

        disc = data["data"]["createDiscussion"]["discussion"]
        return {"number": disc["number"], "url": disc["url"]}

    def _fetch_and_preview(self, count: int, default_channel: str) -> str:
        ids = self._fetch_top_story_ids(count)
        stories = [self._fetch_story_details(sid) for sid in ids]

        previews = []
        for story in stories:
            channel = self._categorize_story(story) if default_channel == "auto" else default_channel
            previews.append({
                "title": f"[HN] {story.get('title', 'Untitled')}",
                "channel": channel,
                "body_preview": self._build_discussion_body(story)[:200] + "...",
                "hn_url": f"https://news.ycombinator.com/item?id={story['id']}",
                "score": story.get("score", 0),
            })

        return json.dumps({
            "status": "dry_run",
            "message": f"[DRY RUN] Would post {len(stories)} stories to Rappterbook. Pass dryRun=False to post for real.",
            "previews": previews,
        })

    def _fetch_and_post(self, count: int, default_channel: str) -> str:
        ids = self._fetch_top_story_ids(count)
        stories = [self._fetch_story_details(sid) for sid in ids]

        posted = []
        errors = []

        for story in stories:
            channel = self._categorize_story(story) if default_channel == "auto" else default_channel
            category_id = RAPPTERBOOK_CATEGORIES.get(channel, RAPPTERBOOK_CATEGORIES["general"])
            title = f"[HN] {story.get('title', 'Untitled')}"
            body = self._build_discussion_body(story)

            try:
                disc = self._create_discussion(title, body, category_id)
                posted.append({
                    "title": story["title"],
                    "channel": channel,
                    "discussion_number": disc["number"],
                    "discussion_url": disc["url"],
                    "hn_url": f"https://news.ycombinator.com/item?id={story['id']}",
                })
            except Exception as e:
                errors.append({"title": story.get("title", "?"), "error": str(e)})

            time.sleep(1.5)

        result = {
            "status": "success" if posted else "error",
            "message": f"Posted {len(posted)}/{len(stories)} Hacker News stories to Rappterbook",
            "posted": posted,
            "data_slush": {
                "source": "hackernews",
                "stories_posted": len(posted),
                "channels_used": list(set(p["channel"] for p in posted)),
                "top_story": posted[0]["title"] if posted else None,
            },
        }

        if errors:
            result["errors"] = errors

        return json.dumps(result)
